kb/development/script-coding-standards.md

```diff
@@ -0,0 +1,102 @@
+# Mbed TLS script coding standards
+
+## Choice of scripting language
+
+The preferred language for scripts in the Mbed TLS project is Python. Use Python for new scripts that are committed to the project unless there's a good reason.
+
+Some reasons not to use Python:
+
+* A script that only applies to Windows and should ideally work out of the box can be written in cmd.
+* Some CI code in the `mbedtls-test` repository is written in Groovy, for its Jenkins integration.
+* A script only intended for the Mbed TLS developers, and whose job is mostly to run other programs, can be written in sh.
+* If you're extending existing scripts, then obviously keep using the same language.
+* Throwaway scripts that do ad hoc automated refactoring (e.g. renaming identifiers) are often in Perl, thanks to its convenient text manipulation primitives.
+
+### Language versions
+
+Check [`README.md`](https://github.com/Mbed-TLS/mbedtls/blob/development/README.md#tool-versions) for the current minimum supported version of Python and other tools. Note that in addition to the public commitments, at any given time, there may be additional constraints due to the platforms on which we run automated tests. As of September 2022, Python scripts must be compatible with Python 3.5.
+
+For scripts that need to be backported, note that [long-time support branches](https://github.com/Mbed-TLS/mbedtls/blob/development/BRANCHES.md#long-time-support-branches) may support older versions of Python and other tools than the `development` branch.
```

Comment:
Type annotations were added to python in version 3.5, so those scripts wouldn't run on anything older than that.
(Comment-style annotations are compatible with earlier python versions as well.)

I think we implicitly dropped support for older version of python in the development (2.x) branch when we stopped testing on Ubuntu 14.04, which ended LTS support in 2019.

Reply:
`config.py`, which is the one script that ordinary users need, still appears to work under Python 3.4. It's from before we added type annotations. Many other scripts require at least Python 3.5 just to parse the type annotations.

We never had CI with Python 3.4: when we switched from Python 2 to Python 3, our CI was Ubuntu 16.04 only (plus more recent Python on FreeBSD and Windows).

I think it would make sense to preserve backward compatibility with Python 3.4 for `config.py` because it's essential. The other scripts are for testing or for contributors, and we can require a more recent version even in the LTS branch.

I think it's better if there's a single point of truth for minimal versions, in individual branches's README, rather than duplicate the information in the coding standards and predictably forget to update it. The only reason I mentioned 3.5 here is that the requirement comes from the current CI, it's not something you can deduce from reading the code in the mbedtls repository.

Reply:
Agree about specifying versions in the branches themselves (which we have started doing for development and perhaps should add to 2.28). If config.py is the solitary exception, then we should note that in a comment at the top of config.py.

Reply:
I've reorganized and reformulated the Python version requirements. I added an explicit mention of 3.4 for `config.py` here: it is going to remain stable for another couple of years. I'll make a PR on development and 2.28 to also mention it in `config.py` itself.